File: parallelPCA.go
package main

import (
	"fmt"
	"math"
	"math/rand"
	"runtime"
	"sync"
	"time"

	"gonum.org/v1/gonum/mat"
	"gonum.org/v1/plot"
	"gonum.org/v1/plot/plotter"
	"gonum.org/v1/plot/vg"
)

// generateDataset creates a synthetic dataset
func generateDataset(nSamples, nFeatures int) *mat.Dense {
	data := make([]float64, nSamples*nFeatures)
	for i := range data {
		data[i] = rand.Float64()
	}
	return mat.NewDense(nSamples, nFeatures, data)
}

// computeCovariance calculates the covariance matrix
func computeCovariance(X *mat.Dense) *mat.SymDense {
	var centered mat.Dense
	centered.CloneFrom(X)
	nSamples, nFeatures := X.Dims()

	for j := 0; j < nFeatures; j++ {
		col := mat.Col(nil, j, X)
		mean := mat.Sum(mat.NewVecDense(len(col), col)) / float64(nSamples)
		for i := 0; i < nSamples; i++ {
			centered.Set(i, j, X.At(i, j)-mean)
		}
	}

	var cov mat.SymDense
	cov.SymOuterK(1/(float64(nSamples)-1), &centered)
	return &cov
}

// eigenDecomposition performs eigen decomposition on the covariance matrix
func eigenDecomposition(cov *mat.SymDense) ([]float64, *mat.Dense) {
	var eigen mat.Eigen
	ok := eigen.Factorize(cov, mat.EigenRight)
	if !ok {
		panic("Eigendecomposition failed")
	}

	values := eigen.Values(nil)
	vectors := eigen.Vectors()

	// Sort eigenvalues and eigenvectors
	n := len(values)
	indices := make([]int, n)
	for i := range indices {
		indices[i] = i
	}
	for i := 0; i < n-1; i++ {
		for j := i + 1; j < n; j++ {
			if math.Abs(values[indices[i]]) < math.Abs(values[indices[j]]) {
				indices[i], indices[j] = indices[j], indices[i]
			}
		}
	}

	sortedValues := make([]float64, n)
	sortedVectors := mat.NewDense(n, n, nil)
	for i, idx := range indices {
		sortedValues[i] = values[idx]
		col := mat.Col(nil, idx, vectors)
		sortedVectors.SetCol(i, col)
	}

	return sortedValues, sortedVectors
}

// timePCA measures the execution time of PCA
func timePCA(numThreads int, X *mat.Dense, numRuns int) float64 {
	runtime.GOMAXPROCS(numThreads)
	totalTime := 0.0

	for i := 0; i < numRuns; i++ {
		start := time.Now()

		cov := computeCovariance(X)

		var wg sync.WaitGroup
		wg.Add(1)
		go func() {
			defer wg.Done()
			eigenDecomposition(cov)
		}()
		wg.Wait()

		elapsed := time.Since(start).Seconds()
		totalTime += elapsed
	}

	return totalTime / float64(numRuns)
}

func main() {
	datasetConfigs := [][2]int{
		{1000, 50}, {1000, 500}, {1000, 5000},
		{5000, 50}, {5000, 500}, {5000, 5000},
		{10000, 50}, {10000, 500}, {10000, 5000},
	}

	threadCounts := []int{1, 2, 4, 8, 16, 32, 64}
	results := make(map[[2]int]map[string][]plotter.XYs)

	for _, config := range datasetConfigs {
		nSamples, nFeatures := config[0], config[1]
		fmt.Printf("Processing dataset with %d samples and %d features...\n", nSamples, nFeatures)

		X := generateDataset(nSamples, nFeatures)

		executionTimes := make(plotter.XYs, len(threadCounts))
		speedups := make(plotter.XYs, len(threadCounts))

		singleThreadTime := timePCA(1, X, 5)

		for i, numThreads := range threadCounts {
			execTime := timePCA(numThreads, X, 5)
			speedup := singleThreadTime / execTime

			executionTimes[i].X = float64(numThreads)
			executionTimes[i].Y = execTime
			speedups[i].X = float64(numThreads)
			speedups[i].Y = speedup
		}

		results[config] = map[string][]plotter.XYs{
			"execution_times": {executionTimes},
			"speedups":        {speedups},
		}
	}

	// Plotting results
	for _, config := range datasetConfigs {
		nSamples, nFeatures := config[0], config[1]

		p, err := plot.New()
		if err != nil {
			panic(err)
		}

		p.Title.Text = fmt.Sprintf("Execution Time (%d samples x %d features)", nSamples, nFeatures)
		p.X.Label.Text = "Number of Threads"
		p.Y.Label.Text = "Execution Time (seconds)"

		err = plotutil.AddLinePoints(p, "Execution Time", results[config]["execution_times"][0])
		if err != nil {
			panic(err)
		}

		if err := p.Save(4*vg.Inch, 4*vg.Inch, fmt.Sprintf("execution_time_%d_%d.png", nSamples, nFeatures)); err != nil {
			panic(err)
		}

		p, err = plot.New()
		if err != nil {
			panic(err)
		}

		p.Title.Text = fmt.Sprintf("Speedup (%d samples x %d features)", nSamples, nFeatures)
		p.X.Label.Text = "Number of Threads"
		p.Y.Label.Text = "Speedup"

		err = plotutil.AddLinePoints(p, "Speedup", results[config]["speedups"][0])
		if err != nil {
			panic(err)
		}

		if err := p.Save(4*vg.Inch, 4*vg.Inch, fmt.Sprintf("speedup_%d_%d.png", nSamples, nFeatures)); err != nil {
			panic(err)
		}
	}
}

File: parallelPCA.py
import numpy as np
from concurrent.futures import ThreadPoolExecutor
import time
import matplotlib.pyplot as plt
from sklearn.preprocessing import StandardScaler

# Function to generate synthetic datasets with varying dimensions and sample sizes
def generate_dataset(n_samples, n_features):
    return np.random.rand(n_samples, n_features)

# Compute covariance matrix
def compute_covariance(X):
    return np.dot(X.T, X) / (X.shape[0] - 1)

# Eigen decomposition
def eigen_decomposition(cov_matrix):
    eig_vals, eig_vecs = np.linalg.eigh(cov_matrix)
    sorted_indices = np.argsort(eig_vals)[::-1]
    return eig_vals[sorted_indices], eig_vecs[:, sorted_indices]

# Time PCA with a given number of threads
def time_pca(num_threads, X_std, num_runs=5):
    total_time = 0
    for _ in range(num_runs):
        start_time = time.perf_counter()
        cov_matrix = compute_covariance(X_std)
        with ThreadPoolExecutor(max_workers=num_threads) as executor:
            eig_vals, eig_vecs = executor.submit(eigen_decomposition, cov_matrix).result()
        end_time = time.perf_counter()
        total_time += (end_time - start_time)
    return total_time / num_runs

if __name__ == "__main__":
    # Dataset configurations: (number of samples, number of features)
    dataset_configs = [
        (1000, 50),
        (1000, 500),
        (1000, 5000),
        (5000, 50),
        (5000, 500),
        (5000, 5000),
        (10000, 50),
        (10000, 500),
        (10000, 5000)
    ]

    thread_counts = [1, 2, 4, 8, 12, 16 ,24, 32 ,40 ,56 ,64]
    results = {}

    for n_samples, n_features in dataset_configs:
        print(f"Processing dataset with {n_samples} samples and {n_features} features...")
        X = generate_dataset(n_samples, n_features)

        # Standardize the dataset
        scaler = StandardScaler()
        X_std = scaler.fit_transform(X)

        # Measure execution times for different thread counts
        thread_execution_times = [
            (num_threads, time_pca(num_threads, X_std))
            for num_threads in thread_counts
        ]

        single_thread_time = thread_execution_times[0][1]
        speedup = [
            (num_threads, single_thread_time / exec_time if exec_time > 0 else 0)
            for num_threads, exec_time in thread_execution_times
        ]

        results[(n_samples, n_features)] = {
            "execution_times": thread_execution_times,
            "speedups": speedup,
        }

    # Plot results in one image with legends
    plt.figure(figsize=(14, 12))

    # Plot Execution Time
    plt.subplot(2, 1, 1)
    for ((n_samples, n_features), data) in results.items():
        threads, times = zip(*data["execution_times"])
        plt.plot(threads, times, marker="o", linestyle="-", label=f"{n_samples}x{n_features}")

    plt.title("Execution Time vs Number of Threads")
    plt.xlabel("Number of Threads")
    plt.ylabel("Execution Time (seconds)")
    plt.grid(True)
    plt.legend(title="Dataset Size")
    
    # Plot Speedup
    plt.subplot(2, 1, 2)
    for ((n_samples, n_features), data) in results.items():
        threads, _ = zip(*data["execution_times"])
        _, speedups = zip(*data["speedups"])
        plt.plot(threads, speedups, marker="o", linestyle="-", label=f"{n_samples}x{n_features}")

    plt.title("Speedup vs Number of Threads")
    plt.xlabel("Number of Threads")
    plt.ylabel("Speedup")
    plt.grid(True)
    plt.legend(title="Dataset Size")

    plt.tight_layout()
    plt.savefig("pca_performance_comparison.png")
    plt.show()
